security: Update security

barakmich opened the pull request:
Still working on some of the client improvements, but here's the PR for fixing the list on #2384 

Notably, prefixes are simplified, enabling security is separate from a root user, there's a root role, there's guest access, and obviates some of the "is security enabled" bugs in the decoupling.

xiang90 commented:
Empty users list returns 

```
{"errorCode":100,"message":"Key not found","cause":"/2","index":3}
```

I expect it to return an empty user json collection.

xiang90 commented:
different behavior between `users` and `users`

```
curl http://127.0.0.1:4001/v2/security/users/
{"message":"Invalid path"}
```

```
curl http://127.0.0.1:4001/v2/security/users/
{"errorCode":100,"message":"Key not found","cause":"/2","index":3}
```

yichengq commented:
The commit titles are not helpful. 1. it doesn't indicate the package/submodule the changes are making on 2. it has WIP in it

xiang90 commented:
better ux when the user is not found

currently it returns the raw error from store
```
curl http://127.0.0.1:4001/v2/security/users/foo
{"errorCode":100,"message":"Key not found","cause":"/2","index":3}
```

I expect this return {"error": "user not found" } or something like that.

xiang90 commented:
Role endpoints has the similar issues:

```
curl http://127.0.0.1:4001/v2/security/roles/foo
{"errorCode":100,"message":"Key not found","cause":"/2","index":3}

curl http://127.0.0.1:4001/v2/security/roles/
curl: (52) Empty reply from server

curl http://127.0.0.1:4001/v2/security/roles
{"errorCode":100,"message":"Key not found","cause":"/2","index":3}
``

xiang90 commented:
The UX of role update could be better

```
// create
curl http://127.0.0.1:4001/v2/security/roles/fleet -XPUT -d '{"role":"fleet","permissions":{"kv":{"read":["/fleet/*"],"write":["/fleet"]}}}'
{"role":"fleet","permissions":{"kv":{"read":["/fleet/*"],"write":["/fleet"]}}}

// try to update write to "/bar", actually we need grant 
// i expect to see, role exists. or role exists, use grant/revoke to update role
curl http://127.0.0.1:4001/v2/security/roles/fleet -XPUT -d '{"role":"fleet","permissions":{"kv":{"read":["/fleet/*"],"write":["/bar"]}}}'
{"role":"fleet","permissions":{"kv":{"read":["/fleet/*"],"write":["/fleet"]}}}

// it did not update as expected
curl http://127.0.0.1:4001/v2/security/roles/fleet
{"role":"fleet","permissions":{"kv":{"read":["/fleet/*"],"write":["/fleet"]}}}

```

xiang90 commented:
We should persist the security. After reboot, the security is disabled on the rebooted machine.

xiang90 commented:
Recursive flag can escape the security checking. We need to check that for both get and watch requests.

barakmich commented:
The enable/disable bit should be fixed now, as well as some other problems

xiang90 commented:
```
curl http://127.0.0.1:4001/v2/security/roles/
curl: (52) Empty reply from server
```

xiang90 commented:
```
curl http://127.0.0.1:4001/v2/security/users/foo
{"errorCode":100,"message":"Key not found","cause":"/2/users/foo","index":6}
```

yichengq commented:
```
unihorn@CoreOS ~/g/s/g/c/etcd> curl http://root:123456@127.0.0.1:4001/v2/security/users/root -XPUT -d '{"user":"root","password":"123"}'
{"user":"root","password":"$2a$10$dNQbOoBzPpESVvfsA0rEJuUlatzqX5pEmA0QYl7.oQU0Zbe48CQmW","roles":["root"]}
unihorn@CoreOS ~/g/s/g/c/etcd> curl http://root:123456@127.0.0.1:4001/v2/security/users/root -XPUT -d '{"user":"root","password":"123"}'
{"user":"root","password":"$2a$10$Pd6zJJEpUpf1hWwjjWLmAeDZeohjruOGDeVPrwOH5mCOCjggv3JOS","roles":["root"]}
unihorn@CoreOS ~/g/s/g/c/etcd> curl http://root:123456@127.0.0.1:4001/v2/security/users/root -XPUT -d '{"user":"root","password":"123"}'
{"user":"root","password":"$2a$10$Si0YCPuLlfVh6fz8fZyMbOTJyN/rzTBIz2.peSmnRUVp/oTmTrHWm","roles":["root"]}
```

even if i change the password, i can still change the root password using old password.
And the hashed password is different each time.

yichengq commented:
```
$ curl http://root:root@127.0.0.1:4001/v2/security/users/root -XDELETE
$ curl http://root:root@127.0.0.1:4001/v2/security/users
{"message":"Insufficient credentials"}
```

yichengq commented:
```
$ curl http://root:123@127.0.0.1:4001/v2/security/enable -XPUT
$ curl http://127.0.0.1:4001/v2/keys/foo
{"message":"Insufficient credentials"}
```
This violates `add everyone user and role when enabling security, which has all access to key space. the no user:pass falls into everyone.`

yichengq commented:
What is the correct way to set a role? I tried '/users/xxx/grant' and '/users/xxx' with roles set in user struct, but they don't work:
```
$ curl http://root:123456@127.0.0.1:4001/v2/security/users/yichengq/grant -XPUT -d '{"grant":"rocket"}'
{"message":"Invalid path"}
$ curl http://root:123456@127.0.0.1:4001/v2/security/users/yichengq -XPUT -d '{"user":"yichengq","password":"123","roles":["rocket"]}'
{"user":"yichengq","password":"$2a$10$/zAU80ExppCHyeXREAMHIOf942FOYnXo/rfM19NFCiApqxytuw2qW","roles":[]}
```

xiang90 commented:
@barakmich Can we add some test code for security?

```
ok  	github.com/coreos/etcd/etcdserver/security	4.571s	coverage: 42.9% of statements
```

```
github.com/coreos/etcd/etcdserver/etcdhttp/client.go (78.5%)
				
github.com/coreos/etcd/etcdserver/etcdhttp/client_security.go (3.2%)
				
github.com/coreos/etcd/etcdserver/etcdhttp/http.go (87.5%)
				
github.com/coreos/etcd/etcdserver/etcdhttp/peer.go (94.4%)
```

xiang90 commented:
@barakmich Nice. Most my issues have been resolved, except
```
curl http://127.0.0.1:4001/v2/security/users/foo
{"errorCode":100,"message":"Key not found","cause":"/2/users/foo","index":6}

curl http://127.0.0.1:4001/v2/security/roles
{"errorCode":100,"message":"Key not found","cause":"/2","index":3}
```

yichengq commented:
```
unihorn@CoreOS ~/g/src> curl http://127.0.0.1:4001/v2/security/users/root -XPUT -d '{"user":"root","password":"123456"}'
{"user":"root","password":"$2a$10$V3PG58N4zuTyGQ6ibX5sje1RUowIbhmTnWyi8yy40cnkA.4uSMkmm","roles":null}
unihorn@CoreOS ~/g/src> curl http://127.0.0.1:4001/v2/security/users/root -XPUT -d '{"user":"root","password":"123456"}'
{"user":"root","password":"$2a$10$TVXBuJib4cvhsSxvILIABuohYnzz1chmTan1vT2Rrk2gFCG.WuLtq","roles":["root"]}
unihorn@CoreOS ~/g/src> curl http://127.0.0.1:4001/v2/security/users/root -XPUT -d '{"user":"root","password":"123456"}'
{"user":"root","password":"$2a$10$i//cGw/7m.hLAFE2XE4/C.3eKTd5jkg7BSglndc.ogSoYXN7NtIyO","roles":["root"]}
```
The hashed password is always different. Is this intentional?
My thought is that hashed password is used to confirm that your password is set correctly, so it should return the same hash value for the same password.

yichengq commented:
better logging needed:
```
unihorn@CoreOS ~/g/src> curl http://root:12@127.0.0.1:4001/v2/security/users/root -XDELETE
{"message":"security: Cannot delete root user while security is enabled.%!(EXTRA string=root)"}%
```
(Great that root cannot be deleted now!)

yichengq commented:
```
$ curl http://root:123@127.0.0.1:4001/v2/security/enable -XPUT
$ curl http://127.0.0.1:4001/v2/keys/foo
{"message":"Insufficient credentials"}
```
This violates `add everyone user and role when enabling security, which has all access to key space. the no user:pass falls into everyone.`
This provides better UX because the enable doesn't breaking existing clients, and user has to change guest user explicitly to break it.

yichengq commented:
```
unihorn@CoreOS ~/g/src> curl http://root:12@127.0.0.1:4001/v2/security/roles/rocket -XPUT -d '{"role":"rocket","grant":{"kv":{"read":["/abc*"],"write":null}}}'
{"role":"rocket","permissions":{"kv":{"read":["/","/abc","/abc*"],"write":[]}}}
unihorn@CoreOS ~/g/src> curl http://root:12@127.0.0.1:4001/v2/security/roles/rocket -XPUT -d '{"role":"rocket","revoke":{"kv":{"read":["/abcd"],"write":null}}}'
{"message":"security: Role not updated. Use Grant/Revoke to update the role."}
```
It fails because the revoked path does not exist. `Use Grant/Revoke to update the role.` is misleading IMO

yichengq commented:
What is supposed to be the correct way to update a user's roles?
```
unihorn@CoreOS ~/g/src> curl http://root:12@127.0.0.1:4001/v2/security/users/yichengq -XPUT -d '{"user":"yichengq","roles":["root"]}' 
{"message":"security: User not updated. Use Grant/Revoke/Password to update the user."}%                                                                                                                    unihorn@CoreOS ~/g/src> curl http://root:12@127.0.0.1:4001/v2/security/users/yichengq/grant -XPUT -d '{"user":"yichengq","roles":["root"]}'
{"message":"Invalid path"}%                                                                                                                                                                                 unihorn@CoreOS ~/g/src> curl http://root:12@127.0.0.1:4001/v2/security/users/yichengq/grant -XPUT -d '{"user":"yichengq","grantRoles":["root"]}'
{"message":"Invalid path"}%
```
all these three failed.

yichengq commented:
```
unihorn@CoreOS ~/g/src> curl http://xiang:12@127.0.0.1:4001/v2/security/roles/root
{"role":"root","permissions":{"kv":{"read":null,"write":null}}}
```
It may be confusing to display root roles' read and write is null.

philips commented:
We need to add some Documentation and move the existing doc from Documentation/rfc to Documentation. It would be ideal to have some sort of guide that uses the etcdctl stuff too.

philips commented on etcdserver/etcdhttp/client_security.go:
```diff
@@ -51,22 +51,25 @@ func hasRootAccess(sec *security.Store, r *http.Request) bool {
+		}
 	}
-	return ok
+	log.Printf("security: Attempting to use user %s for resource that requires %s role.", username, security.RootRoleName)
```
Perhaps: "security: User %s does not have the %s role which is required to access %s resource"?

philips commented on etcdserver/etcdhttp/client_security.go:
```diff
@@ -88,23 +91,34 @@ func hasKeyPrefixAccess(sec *security.Store, r *http.Request, key string) bool {
+	if role.HasKeyAccess(key, writeAccess) {
+		return true
+	}
+	log.Printf("security: Invalid access for unauthenticated user on key %s.", key)
```
for consistency: in the hasRootAccess we say resource, in this message we say key.

philips commented on etcdserver/security/security.go:
```diff
@@ -31,8 +35,30 @@ import (
 	// StorePermsPrefix is the internal prefix of the storage layer dedicated to storing user data.
 	StorePermsPrefix = "/2"
+	RootRoleName     = "root"
+	GuestRoleName    = "guest"
```
Guest role took me a second to figure out, might be worth mentioning in a comment that this is the "unauthenticated" role

yichengq commented:
Most of them are fixed. A few are still in doubt:
1. 
```
unihorn@CoreOS ~/g/src> curl http://127.0.0.1:4001/v2/security/users/root -XPUT -d '{"user":"root","password":"123456"}'
{"user":"root","password":"$2a$10$V3PG58N4zuTyGQ6ibX5sje1RUowIbhmTnWyi8yy40cnkA.4uSMkmm","roles":null}
unihorn@CoreOS ~/g/src> curl http://127.0.0.1:4001/v2/security/users/root -XPUT -d '{"user":"root","password":"123456"}'
{"user":"root","password":"$2a$10$TVXBuJib4cvhsSxvILIABuohYnzz1chmTan1vT2Rrk2gFCG.WuLtq","roles":["root"]}
unihorn@CoreOS ~/g/src> curl http://127.0.0.1:4001/v2/security/users/root -XPUT -d '{"user":"root","password":"123456"}'
{"user":"root","password":"$2a$10$i//cGw/7m.hLAFE2XE4/C.3eKTd5jkg7BSglndc.ogSoYXN7NtIyO","roles":["root"]}
```
The hashed password is always different. Is this intentional?
My thought is that hashed password is used to confirm that your password is set correctly, so it should return the same hash value for the same password.
2.
I see guest role is working, and no user:pass falls to guest role. And if i delete guest role, no user:pass cannot touch key space. pretty nice.
A better solution that I can see is to add guest user, because user, not role, is the one that reads/writes keys.

barakmich commented:
Hashed password always being different is intentional. bcrypt is awesome.
The api endpoints (`/grant` `/revoke`) are gone, and replaced by PUTting the user struct as an update. The documentation has been updated to reflect this.

Fixed Philips' nits.

barakmich commented:
Fixes #2384 -- any remaining issues should be filed separately

yichengq commented on Documentation/security_api.md:
```diff
@@ -10,37 +10,32 @@ There are three types of resources in etcd
+A user is an identity to be authenticated. Each user can have multiple roles. The user has a capability (such as reading or writing) on the resource if one of the roles has that capability.
 
-The special static `root` user has a ROOT role. (Caps for visual aid throughout)
+A user named `root` is required before security can be enabled. The ROOT role can be granted to multiple users, but `root` is required for recovery purposes.
```
A user named `root` is required before security can be enabled, and it is born to have the ROOT role.

yichengq commented on Documentation/security_api.md:
```diff
@@ -10,37 +10,32 @@ There are three types of resources in etcd
+#### Roles
+Each role has exact one associated Permission List. An permission list exists for each permission on key-value resources. 
+
+The special static ROOT (named `root`) role has a full permissions on all key-value resources, the permission to manage user resources and settings resources. Only the ROOT role has the permission to manage user resources and modify settings resources.
```
Does the root role exist from very beginning or when security is enabled?

barakmich replied:
In order to enable security, root must be created. The error message is helpful.

yichengq commented:
lgtm
I browsed over the code roughly and it seems good.
I tested the cases we describe in #2384 and they are no longer there.
It is a good start for next-round user feedback. :)
